provider/bidengine/order_test.go

```diff
@@ -0,0 +1,434 @@
+		ID:      leaseID,
+		Price:   testutil.AkashCoin(t, 1),
+	}
+
+	subscriber, err := scaffold.testBus.Subscribe()
+	require.NoError(t, err)
+	err = scaffold.testBus.Publish(ev)
+	require.NoError(t, err)
+
+	// Wait for the event to be published
+	<-subscriber.Events()
+
+	// Should not have called unreserve yet
+	scaffold.cluster.AssertNotCalled(t, "Unreserve", mock.Anything, mock.Anything)
+
+	// Shutdown after the message has been published
+	order.lc.Shutdown(nil)
+
+	// Should have called unreserve
+	scaffold.cluster.AssertCalled(t, "Unreserve", scaffold.orderID, mock.Anything)
+}
+
+func Test_ShouldNotBidWhenAlreadySet(t *testing.T) {
+
+	bid := &mtypes.Bid{}
+	order, scaffold := makeOrderForTest(t, bid, nil)
+
+	// Wait for a reserve call
+	<-scaffold.reserveCallNotify
+
+	// Should have called reserve once
+	scaffold.cluster.AssertCalled(t, "Reserve", scaffold.orderID, mock.Anything)
+
+	// Should not have called unreserve
+	scaffold.cluster.AssertNotCalled(t, "Unreserve", mock.Anything, mock.Anything)
+
+	ev := mtypes.EventOrderClosed{
+		Context: sdkutil.BaseModuleEvent{},
+		ID:      scaffold.orderID,
+	}
+
+	err := scaffold.testBus.Publish(ev)
+	require.NoError(t, err)
+
+	// Wait for it to stop
+	<-order.lc.Done()
+
+	// Should have called unreserve during shutdown
+	scaffold.cluster.AssertCalled(t, "Unreserve", scaffold.orderID, mock.Anything)
+
+	var broadcast sdk.Msg
+	select {
+	case broadcast = <-scaffold.broadcasts:
+	default:
+	}
+	// Should have broadcast
+	require.NotNil(t, broadcast)
+	closeBid, ok := broadcast.(*mtypes.MsgCloseBid)
+	require.True(t, ok)
+	require.NotNil(t, closeBid)
+
+	require.Equal(t, closeBid.BidID, bid.BidID)
+
+}
+
+func Test_ShouldRecognizeLeaseCreatedIfBiddingIsSkipped(t *testing.T) {
+	bid := &mtypes.Bid{}
+	order, scaffold := makeOrderForTest(t, bid, nil)
+
+	// Wait for a reserve call
+	<-scaffold.reserveCallNotify
+
+	// Should have called reserve once
+	scaffold.cluster.AssertCalled(t, "Reserve", scaffold.orderID, mock.Anything)
+
+	// Should not have called unreserve
+	scaffold.cluster.AssertNotCalled(t, "Unreserve", mock.Anything, mock.Anything)
+
+	leaseID := mtypes.MakeLeaseID(mtypes.MakeBidID(scaffold.orderID, scaffold.testAddr))
+
+	ev := mtypes.EventLeaseCreated{
+		Context: sdkutil.BaseModuleEvent{},
+		ID:      leaseID,
+		Price:   testutil.AkashCoin(t, 1),
+	}
+
+	err := scaffold.testBus.Publish(ev)
+	require.NoError(t, err)
+
+	// Wait for it to stop
+	<-order.lc.Done()
+
+	// Should not have called unreserve during shutdown
+	scaffold.cluster.AssertNotCalled(t, "Unreserve", mock.Anything, mock.Anything)
+
+	var broadcast sdk.Msg
+
+	select {
+	case broadcast = <-scaffold.broadcasts:
+	default:
+	}
+	// Should never have broadcast
+	require.Nil(t, broadcast)
+}
+
+type testBidPricingStrategy int64
+
+func (tbps testBidPricingStrategy) calculatePrice(gspec *dtypes.GroupSpec, _ lifecycle.Lifecycle) (sdk.Coin, error) {
+	return sdk.NewInt64Coin("akt", int64(tbps)), nil
```

Comment:
I think I am just testing that the result get passed back here, when the bid is published as a transaction

Reply:
Ah, I'd still say lets keep it simple and avoid using `akt` since we don't really use that denom internally.

`testutil.CoinDenom` keeps us consistent for tests too!